text/0026-semantic-versioning.md

```diff
@@ -0,0 +1,109 @@
+
+> The buildpack version:
+>
+> - MUST be in the form <X>.<Y>.<Z> where X, Y, and Z are non-negative integers and must not contain leading zeros.
+>    - Each element MUST increase numerically.
+>    - Buildpack authors will define what changes will increment X, Y, and Z.
+
+
+It does not define what changes will increment the Major (`X`), Minor (`Y`) and Patch (`Z`) versions
+of buildpack releases.
+
+This RFC attempts to set guidelines and good practices for Paketo Buildpack authors/maintainers to follow
+when a buildpack release has to be created.
+
+## Motivation
+
+As a Buildpack/Builder maintainer,<br/>
+I need a release versioning guideline,
+so I can make my release versions consistent.
+
+As a Buildpack/Builder user,<br/>
+I want to know how versioning is done,
+so I can use the appropriate version and
+decide when to upgrade.
+
+## Semantics
+
+Semantic Versioning 2.0.0 defines versioning syntax as follows [semver.org](https://semver.org/):
+
+> Given a version number MAJOR.MINOR.PATCH, increment the:
+> 
+>   - MAJOR version when you make incompatible API changes,
+>   - MINOR version when you add functionality in a backwards compatible manner, and
+>   - PATCH version when you make backwards compatible bug fixes.
+
+The proposal is to conform to the above semantics as much as possible.
+
+### Changes
+
+In terms of Paketo Buildpacks/Builders, this translates to:
+
+* Bumping MAJOR is analogous to an "`app or related buildpacks may not work without change`" relation between releases.
+* Bumping MINOR is analogous to an "`updates or adds features`" relation between releases.
+* Bumping PATCH is analogous to a "`fixes issue`" relation between releases.
+
+The following table lists the usual circumstances/events where bumping a particular component is expected:
+
+#### Buildpack
+| `Major` | `Minor` | `Patch` |
+|-|-|-|
+| \<Any incompatible change\> | \<Feature additions but compatible with previous release\> | \<Fixes but no major feature changes\> |
+| Change in Buildpack ID | Addition/removal of dependency<br>(when not superseded by conditions in the Major column) | Bug fixes |
+| Change in supported Stacks | Buildpack runs a new command the causes backward compatible<br>change in behavior (e.g. x-install buildpack adds running "x clean") | Performance improvement, internal tweaks |
+| Change in Buildpack API version | Change of file type of build/detect executable<br>(e.g. text to binary) | CVE fixes |
+| Change in API (Provides/Requires) | New configuration option (e.g. new environment variable support) | Change of required build tool version to package buildpack<br>(e.g. go version) |
+| Removal of configuration options<br>(e.g. buildpack.yml) | Change in <br>[buildpack-specific metadata](https://github.com/buildpacks/spec/blob/buildpack/v0.6/buildpack.md#build-plan-toml) |  |
+| Default version of dependency bumped to incompatible version<br>(e.g. php7 to php8) | Publishing to a new image registry |  |
```

Comment:
I think it would be useful to expand more on how dependency versions will be managed, specifically in the context of languages/compilers and language tooling, **where the dependency is the main thing being provided to the end-user and multiple versions are supported in parallel**.

For example, let's say the Ruby's MRI buildpack currently supports (hypothetically):

* `2.5.9`, `2.5.8`
* `2.6.7`, `2.6.6`, `2.6.5`
* `2.7.3`, `2.7.2`, `2.7.1`

How would the buildpack versioning change in the following cases?

* A new series of patch versions are introduced, adding support for `2.6.8` and `2.7.4`.
* A new minor version is introduced, adding support for `2.8.0`
* A new major version is introduced, adding support for `3.0.1`
* An older patch version is removed, removing support for `2.6.5`
* An older minor version is removed, removing support for all `2.5.X` series.

Here are the same cases with my own thoughts on them, but I'm sure there are things I'm overlooking, so take it with a grain of salt.

* A new series of patch versions are introduced, adding support for `2.6.8` and `2.7.4`.
  * This would probably be safe to release as a patch release, since it should be backwards compatible to the user even if they don't have a specific version set for the project, or if they are set as `2.7.*`.
* A new minor version is introduced, adding support for `2.8.0`
  *  This would probably be safe to release as a patch release IF the new version is not set as default version, since it should be backwards compatible to the user even if they don't have a specific version set for the project, or if they are set as `2.*`.
  * If the new version is the new default, then it should probably be a new minor release.
* A new major version is introduced, adding support for `3.0.1`
  *  This would probably be safe to release as a patch release IF the new version is not set as default version, since it should be backwards compatible to the user even if they don't have a specific version set for the project, or if they are set as `2.*`.
  * If the new version is the new default, then it should probably be a new major release.
* An older patch version is removed, removing support for `2.6.5`
  * This is probably more tricky since it can break an application which is depending on the specific version. In order to be fully compatible with semantic versioning, this might require a major release. That said, it might be OK to remove as a minor release if the expected best practice is to pin projects to `M.N.*` format instead of specific patch level.
* An older minor version is removed, removing support for all `2.5.X` series.
  * Requires major version release.

Similarly, buildpacks that provides toolings, such as Ruby`s bundler, could follow the same approach.

Reply:
> That said, it might be OK to remove as a minor release if the expected best practice is to pin projects to M.N.* format instead of specific patch level.

Is it recommended formally in any of the paketo docs that projects pin to minor instead of patch versions of dependencies? If not, this might be great to document.

Reply:
> A new series of patch versions are introduced, adding support for 2.6.8 and 2.7.4.
> This would probably be safe to release as a patch release, since it should be backwards compatible to the user even if they don't have a specific version set for the project, or if they are set as 2.7.*.

+1

> A new minor version is introduced, adding support for 2.8.0
> This would probably be safe to release as a patch release IF the new version is not set as default version, since it should be backwards compatible to the user even if they don't have a specific version set for the project, or if they are set as 2.*.
> If the new version is the new default, then it should probably be a new minor release.

+1 

> A new major version is introduced, adding support for 3.0.1, 3.0.0
> This would probably be safe to release as a patch release IF the new version is not set as default version, since it should be backwards compatible to the user even if they don't have a specific version set for the project, or if they are set as 2.*.
> If the new version is the new default, then it should probably be a new major release.

+1

> An older patch version is removed, removing support for 2.6.5
> This is probably more tricky since it can break an application which is depending on the specific version. In order to be fully compatible with semantic versioning, this might require a major release. That said, it might be OK to remove as a minor release if the expected best practice is to pin projects to M.N.* format instead of specific patch level.

I think this should be able to be done in a patch release. No one should be pinning to a specific version, that's an anti-pattern we don't want to facilitate or encourage. In short, as an app dev I should care at all when a a patch release is dropped. 

From the author's standpoint, I also worry about this rule bumping minor versions unnecessarily. Because don't keep many patch releases, most new patch release are going to result in an old patch falling off and being removed. If that triggers a minor bump, the minor version is going to go up quick.

> An older minor version is removed, removing support for all 2.5.X series.
> Requires major version release.

My $0.02 is that this would only trigger a minor bump. IMHO, the goal of the build pack is that you're always on the latest release of a major branch. I should be able to say give me Ruby 2 or 3 or Java 8 or 11, with the rest being details. Thus if a new minor comes out, you should get that right away. In addition, the oldest minor should drop off, so we're keeping at most two. That gives a dev the ability to roll back one version if there's a problem, time to fix the problem, then get back up to the latest version. I think if you're following that path, the removal of the N-2 release, should be a non-event. I could almost see it being done in a patch release, but that is probably going too far.

Reply:
>Default version of dependency bumped to incompatible version

I don't really like the word "incompatible" here. Why not just say major? so when the default version of a dependency changes to a new major version, that triggers a major buildpack version bump.

Reply:
> My $0.02 is that this would only trigger a minor bump. IMHO, the goal of the build pack is that you're always on the latest release of a major branch. I should be able to say give me Ruby 2 or 3 or Java 8 or 11, with the rest being details.

Unfortunately, at least in my experience, this is almost never true. Most companies that I'm aware tend to pin to a specific minor version, because there are many minor changes that can lead to a unexpected change in behavior. Specifically, while minor versions don't usually break APIs, it introduces new features, refactoring and implementation changes, as well as changes to the runtime such as GC, or upgrade to linked dependencies, etc. Those can lead to unintended production consequences.

I definitely see this as a pain for the buildpack authors, though. Having to bump major in order to drop a minor release would/is likely going to be painful, but not doing so would lead to a breaking change if the downstream user has pinned a minor version, or if the language's minor release has added a unintentional change in behavior.

Reply:
I added a separate table for dependency updates - so please take a look.

> I don't really like the word "incompatible" here. Why not just say major?

That language was to not assume that upstream dependencies always follow the same versioning semantics. An example would be `pipenv` who uses ` YYYY.MM.DD` based on release date. ([Pipenv version history](https://pipenv.pypa.io/en/latest/changelog/)) The onus is on maintainers to determine if the new version is incompatible.

> An older minor version is removed, removing support for all 2.5.X series.
Requires major version release.

I think this strategy will probably work for something like `mri` that makes a minor release [about once a year](https://www.ruby-lang.org/en/downloads/releases/), but not for something like `node` that makes [multiple minor bumps a month](https://nodejs.org/en/download/releases/). IMO releasing a large number of major version releases for node could cause confusion and resistance to update among users. But due to this difference, I'd expect node's default version to be constrained to `major` and that of mri to be constrained to `major.minor`. So probably the semantics should be tied to if that removal also causes a change of buildpack's default version.

Reply:
@arjun024 - Table looks good! Thanks

Reply:
@arjun024 that's a really good point. I was mainly thinking of Ruby and Go buildpacks in my use cases. That table looks good to me. Thanks!